### ISSUE TITLE
test_wavefunction failing for ComplexApprox issue with Intel

### ISSUE THREAD

**Issue by ye-luo** [QMCPACK/qmcpack#1871]
https://cdash.qmcpack.org/CDash/testDetails.php?test=6081090&build=77645
```
  REQUIRE( r_all_val == ComplexApprox(ValueType(1.653821746120792, 0.5484992491019633)) )
with expansion:
  (1.65382,0.548499)
  ==
  ComplexApprox( (0.548499,0.548499) )
```
The code is actually correct since the left side matches the reference value on the REQUIRE line.

A much simpler reproducer.
The following check works with RealType as double but not as float
```REQUIRE(std::complex<RealType>(1, 0) == ComplexApprox(std::complex<RealType>(1, 0)));```
The problem only occurs with Intel compiler. GNU and Clang work fine.

Since ComplexApprox is not part of catch, we probably need to have a look more carefully.
@markdewing Could you have a look since you contributed ComplexApprox?
I think we need tests for ComplexApprox. It is part of QMCPACK not Catch. If we have unit tests rely on it, itself must be tested. We don't have any unit tests for it so far.

**Comment by markdewing** [comment]
It appears to be a bug in the Intel compiler (unless the implicit conversion rules for std::complex do unexpected things that I do not understand, which is always a possibility with C++). I posted the issue to the Intel compiler forum (the post might not appear right away):
https://software.intel.com/en-us/forums/intel-c-compiler/topic/820166

The workaround will be to make a ComplexApprox constructor that accepts a std::complex<float> to avoid any conversions.

**Comment by ye-luo** [comment]
I tried to workaround the issue and put
```
    ComplexApprox(std::complex<float> value) : m_value(std::complex<double>(value.real(), value.imag())), m_compare_real_only(false) {
      // Copied from catch.hpp - would be better to copy it from Approx object
      m_epsilon = std::numeric_limits<float>::epsilon()*100;
    }
```
The compiler get confused which implicit conversion should pick.
```
            function "Catch::ComplexApprox::ComplexApprox(std::complex<double>)"
            function "Catch::ComplexApprox::ComplexApprox(std::complex<float>)"
            argument types are: (double)
    REQUIRE(ddpsiV[1](1, 1) == ComplexApprox(-0.07996207476).epsilon(eps).compare_real_only());
```
On the other hand, I don't understand the logic of `compare_real_only()`. If comparing only the real part is intended. We can just do
```
    REQUIRE(std::real(ddpsiV[1](1, 1)) == Approx(-0.07996207476).epsilon(eps));
```

**Comment by ye-luo** [comment]
The second part has been addressed by #1935
